Store a `MirSource` inside every `Body`

ecstatic-morse opened the pull request:
Resolves #77427.

r? @ghost

ecstatic-morse commented:
@lcnr I don't understand how to use `WithOptConstParam`. What's going wrong here?

ecstatic-morse commented:
r? @lcnr I guess. I was expecting this to be a straightforward refactoring, but #74113 has made me a bit frustrated

It feels like we could make `WithOptConstParam` less invasive when it comes to the `*_mir` queries. A MIR body either comes from an `AnonConst` or not, and that doesn't change as we progress through the optimization pipeline, so why does this type infect everything?

lcnr commented on compiler/rustc_mir_build/src/build/mod.rs:
```diff
@@ -740,6 +750,7 @@ fn construct_error<'a, 'tcx>(hir: Cx<'a, 'tcx>, body_id: hir::BodyId) -> Body<'t
 impl<'a, 'tcx> Builder<'a, 'tcx> {
     fn new(
         hir: Cx<'a, 'tcx>,
+        def_id: DefId,
```
```suggestion
        def: ty::WithOptConstParam<DefId>,
```
I think it should be easier to already pass the correct `WithOptConstParam` into this builder.

ecstatic-morse replied:
That's the part I'm not understanding. What are the invariants of `WithOptConstParam`? When is it okay to be `unknown`? Do I have to call `try_upgrade` manually?

lcnr replied:
The invariant of `WithOptConstParam` is that if `self.const_param_did.is_some()`, then `self.const_param_did == tcx.opt_const_param_of(self.did)`. To prevent cycle errors `self.const_param_did` should also be `Some` if it is used during `typeck`, `did` is a const argument and we want to know `tcx.type_of(did)`, which we then replace with `tcx.type_of(def.def_id_for_type_of())`.

You should never have to call `try_upgrade` except at the start of a query taking `ty::WithOptConstParam` so that we only evaluate it once.

It is okay to use `unknown` if we are either done with typeck or `tcx.opt_const_param_of(self.did)` would/did return `None`.

Looks like it might be worth it to add a section to the rustc-dev-guide about `WithOptConstParam` :thinking:

ecstatic-morse replied:
> Looks like it might be worth it to add a section to the rustc-dev-guide about WithOptConstParam thinking

As I've alluded to, I think time would be better spent better encapsulating this system. I don't particularly want to think about it when writing MIR transformations. I realize I'm being a bit curmudgeonly here and that this is in service of a feature which I'm super excited about (thanks BTW!), but technical debt is a real concern, especially in mir-opt land.

Is there anything we could do to reduce the surface area here? We can discuss this in a separate issue I guess.

lcnr replied:
Looks like I didn't send the message I wanted to send here :sweat_smile: 

While working on this I experimented with some different approaches and the less invasive ones all ended up being unsound in the face of incremental compilation. So I would be glad - and a bit annoyed :laughing: - if you are able to come up with a less invasive solution here, but do think that this will be quite hard to do

ecstatic-morse replied:
I have some ideas, but if you've already explored alternatives, it's unlikely that I'll be able to do better. I've [not given up quite yet](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Alternatives.20to.20.60ty.3A.3AWithOptConstParam.60) though :smile:.

lcnr reviewed (commented):
So `WithOptConstParam` is used if we ever call `type_of` for the mir body. 

As we call `type_of` in some annoyingly annoying places, we have to somehow get the `DefId` of the relevant const param there.
Once we build a `Body` with the correct `WithOptConstParam`, we should then be able to use that `mir_source` in the remaining `mir_*` queries.

`WithOptConstParam` infects all of the mir queries as we call `optimized_mir` during typeck (which is where the cycle errors are the problem) so we have to somehome thread these calls down to `mir_built` which uses `type_of`. Not sure if `type_of` is called in some other mir queries

lcnr commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -11,11 +11,12 @@ use crate::ty::fold::{TypeFoldable, TypeFolder, TypeVisitor};
 use crate::ty::subst::{Subst, SubstsRef};
 use crate::ty::{
-    self, AdtDef, CanonicalUserTypeAnnotations, List, Region, Ty, TyCtxt, UserTypeAnnotationIndex,
+    self, AdtDef, CanonicalUserTypeAnnotations, InstanceDef, List, Region, Ty, TyCtxt,
```
nit: maybe split so it fits into one line

lcnr commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -112,6 +113,38 @@ impl MirPhase {
+}
+
+impl<'tcx> MirSource<'tcx> {
+    pub fn item(def_id: DefId) -> Self {
```
```suggestion
    pub fn item(def: ty::WithOptParam<DefId>) -> Self {
```
my personal preference, makes it harder to get this wrong

ecstatic-morse replied:
This isn't mine. I've just moved it from `transform/mod.rs`.

lcnr replied:
yeah, I still think that this would be a good change though :thinking: you don't have to do this in this PR

lcnr commented on compiler/rustc_mir/src/transform/add_moves_for_packed_drops.rs:
```diff
@@ -40,9 +40,9 @@ use crate::util::patch::MirPatch;
-        add_moves_for_packed_drops(tcx, body, src.def_id());
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!("add_moves_for_packed_drops({:?} @ {:?})", body.source, body.span);
+        add_moves_for_packed_drops(tcx, body, body.source.def_id());
```
```suggestion
        add_moves_for_packed_drops(tcx, body);
```
we only use to body source in `add_moves_for_packed_drops`, don't we?

lcnr commented on compiler/rustc_mir/src/transform/remove_unneeded_drops.rs:
```diff
@@ -11,13 +11,13 @@ use super::simplify::simplify_cfg;
             body,
             optimizations: vec![],
-            def_id: source.def_id().expect_local(),
+            def_id: body.source.def_id().expect_local(),
```
```suggestion
```
Can we instead use the `body` def_id here?

bors commented:
:umbrella: The latest upstream changes (presumably #77396) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

ecstatic-morse commented:
There's a follow-up to #77427, which is to remove the `DefId` from all `(DefId, Body)` pairs in function signatures. I'm not going to it as part of this PR (though I think I did a few functions by accident in `pretty.rs`) since it's already prone to merge conflicts.

ecstatic-morse commented on compiler/rustc_mir_build/src/build/mod.rs:
```diff
@@ -29,7 +29,13 @@ crate fn mir_built<'tcx>(
     }
 
-    tcx.alloc_steal_mir(mir_build(tcx, def))
+    let mut body = mir_build(tcx, def);
+    if def.const_param_did.is_some() {
+        assert!(matches!(body.source.instance, ty::InstanceDef::Item(_)));
+        body.source = MirSource::from_instance(ty::InstanceDef::Item(def.to_global()));
+    }
+
+    tcx.alloc_steal_mir(body)
```
@lcnr what should this look like in your eyes? It seems like we should do this when the MIR is built instead, but that means calling `try_upgrade` outside of a query?

lcnr replied:
```rust
let mut body = mir_built(txc, def);
```

lcnr replied:
we already have `def` as an input to `mir_built`, so we should be able to use the correct `source` right from the start here.

ecstatic-morse replied:
That causes cycle errors during type-checking.

lcnr replied:
why do you think that we should call `try_upgrade` here?

ecstatic-morse replied:
Define "here"? There's already a call to `try_upgrade` above.

ecstatic-morse replied:
This commit is a hack that I stumbled on after some trial and error to avoid cycle errors on some const-generics tests. I was hoping to learn what the correct approach would look like.

lcnr replied:
>  I meant inside rustc_mir_builder::Builder::build() or so.

yeah, we should never have to call `try_upgrade` inside of `mir_built` after the initial call at the start of the query. If you can get the failing state again I would gladly look into what's going wrong there

ecstatic-morse replied:
```
error[E0391]: cycle detected when type-checking `main`
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:33
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                 ^^^^^^^^^^^^^
   |
note: ...which requires simplifying constant for the type system `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
note: ...which requires const-evaluating + checking `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
   = note: ...which requires normalizing `main::{constant#9}::promoted[0]`...
note: ...which requires resolving instance `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing the optional const parameter of `main::{constant#9}`...
  --> /home/mackendy/src/rust/rust/alt/src/test/ui/const-generics/slice-const-param-mismatch.rs:18:46
   |
LL |     let _: ConstBytes<b"AAA"> = ConstBytes::<{&[0x41, 0x41, 0x41]}>;
   |                                              ^^^^^^^^^^^^^^^^^^^^^
   = note: ...which again requires type-checking `main`, completing the cycle
   = note: cycle used when type-checking all item bodies

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0391.
For more information about an error, try `rustc --explain E0308`.

------------------------------------------



failures:
    [ui] ui/const-generics/slice-const-param-mismatch.rs#full
    [ui] ui/const-generics/slice-const-param.rs#full
```

Cycle along with the two failing tests. This is what happens with *just* the first commit.

lcnr replied:
```
   = note: ...which requires normalizing `main::{constant#9}::promoted[0]`...
```
Is probably where this goes wrong, so we somehow are not  using the full `def` for promoteds.

Not at home rn, will look at where exactly this goes wrong tomorrow.

lcnr replied:
So your solution is probably the easiest for now...

T the actual mir builder currently only uses a `DefId` and has quite a few entry points.
After we merge this I am going to look into a good way to start with the correct `WithOptConstParam` in the `build::Builder`.
The way I expect to do this is a bit like the following but without using `WithOptConstParam::unknown` there, which is a bit more invasive than I would like in this PR: https://github.com/lcnr/rust/commit/90615386cc3d8a3a7f3edf7bd0428dcc3f987bd3

lcnr replied:
Apart from this I think this is mostly ready for merge :thinking: is there still something you want to do in that PR?

ecstatic-morse commented:
@bors p=1 rollup=never

(prone to merge conflicts)

@lcnr I cleaned up the commit messages and rebased. Nothing else has changed. This should be ready for the queue now.

lcnr commented:
yeah, let's fix the remaining nits in followup PRs

@bors r+

bors commented:
:pushpin: Commit 606655edc431d5f0069543a5d523c8bc67b918cf has been approved by `lcnr`

<!-- @bors r=lcnr 606655edc431d5f0069543a5d523c8bc67b918cf -->
<!-- homu: {"type":"Approved","sha":"606655edc431d5f0069543a5d523c8bc67b918cf","approver":"lcnr"} -->

bors commented:
:hourglass: Testing commit 606655edc431d5f0069543a5d523c8bc67b918cf with merge 4ccf5f731bb71db3470002d6baf5ab4792b821d9...
<!-- homu: {"type":"BuildStarted","head_sha":"606655edc431d5f0069543a5d523c8bc67b918cf","merge_sha":"4ccf5f731bb71db3470002d6baf5ab4792b821d9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1206320896), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37021)
Approved by: lcnr
Pushing 4ccf5f731bb71db3470002d6baf5ab4792b821d9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37021","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1206320896"},"merge_sha":"4ccf5f731bb71db3470002d6baf5ab4792b821d9"} -->

RalfJung commented:
This PR introduced an ICE for mir-opt-level 2 and above: https://github.com/rust-lang/rust/issues/77564.

lcnr commented on compiler/rustc_mir/src/transform/inline.rs:
```diff
@@ -300,7 +301,7 @@ impl Inliner<'tcx> {
         // FIXME: Give a bonus to functions with only a single caller
 
-        let param_env = tcx.param_env(self.source.def_id());
+        let param_env = tcx.param_env(callee_body.source.def_id());
```
@ecstatic-morse this is the wrong source, will open a PR for this

lcnr replied:
opened #77568
